cored/src/commonTest/kotlin/com/github/kittinunf/cored/ReduxWithMapbackedEngineTest.kt

```diff
@@ -173,12 +179,50 @@ class StoreAdapterTest {
         }
 
         assertEquals(200, sideEffectData.value)
+
+        // Test after add new middleware
+        runTest {
+            localStore.dispatch(Decrement(10)) // 200 - 90 = 110
+        }
+        assertEquals(110, sideEffectData.value)
     }
 
     @Test
-    @Ignore
     fun `should invoke middleware until remove`() {
-        TODO("Not implemented yet")
+        data class SideEffectData(var value: Int)
+
+        val sideEffectData = SideEffectData(100)
+
+        val middleware = AnyMiddleware { order: Order, store: CounterStore, state: CounterState, action: Any ->
+            if (order == Order.BeforeReduce) {
+                assertEquals(0, state.counter)
+                assertTrue(action is Increment)
+            } else {
+                sideEffectData.value = sideEffectData.value + state.counter
+            }
+        }
+
+        val localStore = Store(testScope, CounterState(), reducers, emptyMap())
+        localStore.addMiddleware(Increment::class, middleware)
+
+        runTest {
+            localStore.states
+                .withIndex()
+                .printDebug()
+                .launchIn(testScope)
+
+            localStore.dispatch(Increment(100))
+        }
+        assertEquals(200, sideEffectData.value)
+
+        localStore.removeMiddleware(Increment::class, middleware)
+
+        runTest {
+            localStore.dispatch(Increment(100))
+            localStore.dispatch(Decrement(100))
+        }
+
+        assertEquals(200, sideEffectData.value)
```

Comment:
This feels weird 🤔 
So we have:
```kotlin
        assertEquals(200, sideEffectData.value)
```
Then we remove `Increment::class` middleware, so increasing should stop working but decreasing shouldn't be affected, right?

Although later we have increase by 100 and decrease by 100. Assuming increasing doesn't work at this point anymore, our final result shouldn't be 100? (as sideEffectData.value was 200 before and we decreased it by 100)

Reply:
The thing is `sideEffectData.value` only interact with `Increment` Action right?.

Just to summarize the order is this;

-> addMiddleware for Increment Action
```
T1 : Increment(100) -> CounterState(counter=100), SideEffectData(value = 200) // CounterState (0 -> 100), SideEffectData (100->200) (with effect of middleware)

-> removeMiddleware for Increment Action
T2: Increment(100)  -> CounterState(counter=200), SideEffectData(value = 200)  // CounterState (100 -> 200), SideEffectData doesn't change because middleware is removed

T3: Decrement(100) -> CounterState(counter=100), SideEffectData(value = 200) // CounterState (200 -> 100), SideEffectData is also intact
```

Reply:
I add `assertEquals` for counterState so it makes more sense for the reader.

Reply:
Ohh I see now! It was quite confusing with having counter value in the store and `sideEffectData.value` outside which is modified by middleware :P 

[nit] To make it a little bit more readable, maybe adding 
```kotlin
assertEquals(100, localStore.currentState.counter)
```
before you remove the middleware and then do something like:

```kotlin
        runTest {
            localStore.dispatch(Increment(100))
            localStore.dispatch(Increment(100))
            localStore.dispatch(Decrement(100))
        }
``` 

so we can see that the `localStore.currentState.counter` has changed (from 100 -> 200) but `sideEffectData.value` got intact.

Reply:
We are thinking the same, I did! :) https://github.com/kittinunf/CoRed/pull/20/commits/d982840f8437c6decbc5458bed2870007dbc9143#diff-4be08650aab6229fa6f258abc867b283254743ec97c62f1cf5066df50e77739cR225